File: fib.pl
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% fib.pl
% 
% Prolog implementation of fib that runs in linear time.
% Author: Ramin Rakhamimov
% http://raminrakhamimov.com
% ramin32 AT gmail DOT com
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

fib(0,0).
fib(1,1).
fib(2,2).
fib(F,R) :- fib(F,3,1,2,R).

fib(F,F,_,Y,R) :- R is Y, !.
fib(F,I,X,Y,R) :-
  I1 is I + 1,
  X1 is Y,
  Y1 is X + Y,
  fib(F,I1,X1,Y1,R).
  

File: fib.c
/****************************************
 * fib.c
 *
 * Optimized version of recursive fib.
 * Runs in O(n) time.
 *
 * Author:
 * Ramin Rakhamimov
 * ramin32 at gmail dot com
 * http://raminrakhamimov.com
 ****************************************/

#include <stdio.h>

int fib(int);
int fib_helper(int, int, int);

int main(int argc, char** argv)
{
    printf("%d\n", fib(atoi(argv[1])));
}

int fib(int n)
{
    int prev1 = 0;
    int prev2 = 1;
    return fib_helper(n, prev1, prev2);
}

int fib_helper(int n, int prev1, int prev2)
{
    if(n <= 0)
        return prev2;

    return fib_helper(n-1, prev2, prev1 + prev2);
}

